sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,426 @@
+            on_negotiation_needed(&app_state_clone, values, &out_clone)
+        })
+        .unwrap();
+
+    let out_clone = out.clone();
+    let app_state_clone = app_state.clone();
+    webrtc
+        .connect("on-ice-candidate", false, move |values| {
+            send_ice_candidate_message(&app_state_clone, values, &out_clone)
+        })
+        .unwrap();
+
+    let pipe_clone = pipe.clone();
+    // TODO: replace with webrtc.connect_pad_added
+    webrtc
+        .connect("pad-added", false, move |values| {
+            on_incoming_stream(values, &pipe_clone)
+        })
+        .unwrap();
+
+    let ret = pipe.set_state(gst::State::Playing);
+    assert_ne!(ret, gst::StateChangeReturn::Failure);
+
+    return webrtc;
+}
+
+struct WsClient {
+    out: ws::Sender,
+    webrtc: Option<gst::Element>,
+    app_state: Arc<AtomicUsize>,
+}
+
+impl WsClient {
+    fn update_state(&self, state: AppState) {
+        self.app_state.store(state as usize, Ordering::Relaxed);
+    }
+}
+
+impl ws::Handler for WsClient {
+    fn on_open(&mut self, _: ws::Handshake) -> ws::Result<()> {
+        self.update_state(AppState::ServerConnected);
+        // TODO: replace with random
+        self.update_state(register_with_server(&self.app_state, &self.out));
+        Ok(())
+    }
+
+    fn on_message(&mut self, msg: ws::Message) -> ws::Result<()> {
+        // Close the connection when we get a response from the server
+        let msg_text = msg.into_text().unwrap();
+        if msg_text == "HELLO" {
+            if self.app_state.load(Ordering::Relaxed) != (AppState::ServerRegistering as usize) {
+                // TODO: signal and cleanup
+                panic!("ERROR: Received HELLO when not registering");
+            }
+            self.update_state(AppState::ServerRegistered);
+            setup_call(&self.app_state, &self.out);
+            return Ok(());
+        }
+        if msg_text == "SESSION_OK" {
+            if self.app_state.load(Ordering::Relaxed) != (AppState::PeerConnecting as usize) {
+                // TODO: signal and cleanup
+                panic!("ERROR: Received SESSION_OK when not calling");
+            }
+            self.update_state(AppState::PeerConnected);
+            self.webrtc = Some(start_pipeline(&self.app_state, &self.out));
+            return Ok(());
+        }
+
+        if msg_text.starts_with("ERROR") {
+            println!("Got error message! {}", msg_text);
+            let error = match self.app_state.load(Ordering::Relaxed) {
+                x if x == AppState::ServerConnecting as usize => AppState::ServerConnectionError,
+                x if x == AppState::ServerRegistering as usize => AppState::ServerRegisteringError,
+                x if x == AppState::PeerConnecting as usize => AppState::PeerConnectionError,
+                x if x == AppState::PeerConnected as usize => AppState::PeerCallError,
+                x if x == AppState::PeerCallNegotiating as usize => AppState::PeerCallError,
```

Comment:
See above, if you make the app struct a `Rc<Mutex<_>>` then you could store the state directly as an enum and match properly on it, exhaustively

Reply:
Can we get away with `Rc` with the promises? I assumed that would behave the same way as spawning a new thread and a `sync::Arc` would be needed.